### PR TITLE
Fix incorrect curse priority

### LINKED ISSUE #2808: Curse priority is incorrect for multiple corrupted curse implicits
Using items with multiple corrupted curse on hit implicit does not apply the correct curse in the calculations. 

From the [community wiki](https://pathofexile.fandom.com/wiki/Curse):
> With multiple Curse on Hit corruption implicits on the same item (e.g. Hands of the High Templar gloves, double-corrupted items), the earliest alphabetically-named curse has highest priority, up to the curse limit.

This is currently not working as intended for this pair of gloves:
```
Hands of the High Templar
Crusader Gloves
Item Level: 86
Quality: 11
Sockets: W-G-B-R
LevelReq: 66
Implicits: 5
+2 to Level of Socketed AoE Gems
+2 to Level of Socketed Aura Gems
Curse Enemies with Vulnerability on Hit, with 44% increased Effect
Curse Enemies with Elemental Weakness on Hit, with 40% increased Effect
+0.78% to Spell Critical Strike Chance
Armour: 359 (augmented)
Energy Shield: 71 (augmented)
Str: 51
Int: 51
Can be modified while Corrupted
Can have up to 5 Implicit Modifiers while Item has this Modifier
186% increased Armour and Energy Shield
12% increased maximum Life
+26% to Fire and Lightning Resistances
Note: ~b/o 10 exalted
Corrupted
```
ElementalWeakness is first in the alphabet, yet the calculations apply vulnerability instead.

POB: https://pastebin.com/c3aDQ4gS

### PR DESCRIPTION
**Pull request by Lothrik**
Fixes #2808.

### Description of the problem being solved:
Curse on Hit from corruption implicits are always applied after Curse On Hit Support gems, and trigger mark when you hit mods on rings. Curses from rings appear to override curses from gloves. The earliest alphabetically-named curse has highest priority, up to the curse limit. Curses are applied in alphabetical order according to their internal name. Despair's internal name is Vulnerability and Vulnerability's internal name is NewVulnerability.

TL;DR:
- If you have Curse on Hit gloves: the curse from your gloves should always override Hextouch + Elemental Weakness.
- If you have a Frostbite on Hit ring, a Vulnerability on Hit ring, and Elemental Weakness on Hit gloves: Frostbite should override Vulnerability, and Vulnerability should override Elemental Weakness.
- If you have Vulnerability + Elemental Weakness on Hit gloves, Elemental Weakness should override Vulnerability.

Almost everything I just listed above PoB gets wrong in 2.13.0, as it makes no attempt to prioritize curses alphabetically, prioritize curse on hit corruptions over anything, or prioritize ring curse on hit effects over glove curse on hit effects.

I'm not 100% confident about how I'm handling triggered Marks. The wiki claims they're prioritized below implicit Curse on Hit gloves, which makes sense.. sort of.. given that the trigger effect has a cooldown, but I'm suspicious about the exact mechanics involved.

### Link to a build that showcases this PR:
https://pastebin.com/N4eu1SBZ

### CODE REVIEW

File: src/Modules/CalcPerform.lua [hunk 1]
```diff
@@ -1729,7 +1742,7 @@ function calcs.perform(env, avoidCache)
 					local curse = {
 						name = buff.name,
 						fromPlayer = true,
-						priority = activeSkill.skillTypes[SkillType.Aura] and 3 or 1,
+						priority = determineCursePriority(buff.name, activeSkill.socketGroup.slot, activeSkill.socketGroup.source) + (activeSkill.skillTypes[SkillType.Aura] and 30000 or 10000),
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:1745]: Can we change these magic numbers to constants or at least calculate them as part of the `determineCursePriority` function?

File: src/Modules/CalcPerform.lua [hunk 2]
```diff
@@ -642,6 +642,19 @@ local function doActorLifeManaReservation(actor)
 	end
 end
 
+-- Helper function to determine curse priority when processing curses beyond the curse limit
+local function determineCursePriority(curseName, slot, source)
+	local basePriority = data.cursePriority[curseName]
+	local sourcePriority = source ~= nil and 200 or 100
+	local slotPriority = 2000
+	if string.match(curseName, "'s Mark") ~= nil then
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:650]: style

```suggestion
	if curseName:match("'s Mark") then
```

File: src/Modules/CalcPerform.lua [hunk 3]
```diff
@@ -642,6 +642,19 @@ local function doActorLifeManaReservation(actor)
 	end
 end
 
+-- Helper function to determine curse priority when processing curses beyond the curse limit
+local function determineCursePriority(curseName, slot, source)
+	local basePriority = data.cursePriority[curseName]
+	local sourcePriority = source ~= nil and 200 or 100
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:648]:
```suggestion
	local sourcePriority = source and 200 or 100
```

File: src/Modules/CalcPerform.lua [hunk 4]
```diff
@@ -642,6 +642,19 @@ local function doActorLifeManaReservation(actor)
...
+	local slotPriority = 2000
+	if string.match(curseName, "'s Mark") ~= nil then
+		slotPriority = 1000
+	elseif string.match(slot or "", "Ring ") ~= nil then
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:652]: style

```suggestion
	elseif slot:match("Ring %d") then
```

File: src/Modules/CalcPerform.lua [hunk 5]
```diff
@@ -642,6 +642,19 @@ local function doActorLifeManaReservation(actor)
 	end
 end
 
+-- Helper function to determine curse priority when processing curses beyond the curse limit
+local function determineCursePriority(curseName, slot, source)
+	local basePriority = data.cursePriority[curseName]
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:647]:
```suggestion
        slot = slot or ""
	local basePriority = data.cursePriority[curseName]
```

File: src/Modules/CalcPerform.lua [hunk 6]
```diff
@@ -1882,7 +1895,7 @@ function calcs.perform(env, avoidCache)
 					local curse = {
 						name = grantedEffect.name,
 						fromPlayer = (dest == curses),
-						priority = 2,
+						priority = determineCursePriority(buff.name, nil, nil) + 20000,
```
Comment (Wires77) [inline src/Modules/CalcPerform.lua:1898]:
```suggestion
						priority = determineCursePriority(buff.name) + 20000,
```

**Review by Wires77 — CHANGES_REQUESTED** [review]
In hindsight, I should've made the changes as one suggestion, sorry if it's hard to read.  In general, we use `stringVar:match(pattern)` instead of `string.match(stringVar, pattern)` for string functions and `nil` is falsy, so we can use `if var then` vs. `if var ~= nil then`

As a future PR, we might be able to construct the priority table from the export scripts, but this works for now.  I'd like to see a reasonable fallback value if the curse isn't found, though

**Comment by Lothrik** [comment]
All of that should be fixed now, plus some other bugs I noticed while going over this again.

File: src/Modules/Data.lua [hunk 7]
```diff
@@ -252,6 +252,31 @@ data.specialBaseTags = {
...
+	["Conductivity"] = 12,
+	["Assassin's Mark"] = 13,
+	["CurseFromSkillGem"] = 100,
+	["CurseFromEquipment"] = 200,
+	["AnySlotMarkType"] = 1000,
+	["LeftRingSlotMarkType"] = 2000,
+	["RightRingSlotMarkType"] = 3000,
+	["AnySlotHexType"] = 4000,
+	["LeftRingSlotHexType"] = 5000,
+	["RightRingSlotHexType"] = 6000,
```
Comment (Wires77) [inline src/Modules/Data.lua:276]:
```suggestion
	["RightRingSlotMarkType"] = 2000,
	["LeftRingSlotMarkType"] = 3000,
	["AnySlotHexType"] = 4000,
	["RightRingSlotHexType"] = 5000,
	["LeftRingSlotHexType"] = 6000,
```

**Review by Wires77 — APPROVED** [review]
You had left/right swapped, but this looks good now

**Pull request merged**